File: go/problems/search_suggestions_system.go
package problems

import (
	"sort"
)

// https://leetcode.com/problems/search-suggestions-system
func SuggestedProducts(products []string, searchWord string) [][]string {
	sort.Strings(products)
	dest := make([][]string, len(searchWord)+1)
	dest[0] = products
	for i, c := range searchWord {
		for _, s := range dest[i] {
			if len(s) > i && s[i] == byte(c) {
				dest[i+1] = append(dest[i+1], s)
			}
		}
	}
	for i := 1; i < len(dest); i++ {
		if len(dest[i]) > 3 {
			dest[i] = dest[i][:3]
		}
	}
	return dest[1:]
}

File: rust/src/problems/search_suggestions_system.rs
// https://leetcode.com/problems/search-suggestions-system

pub fn suggested_products(mut products: Vec<String>, search_word: String) -> Vec<Vec<String>> {
  let mut results: Vec<Vec<String>> = vec![];
  products.sort();

  search_word
    .chars()
    .into_iter()
    .enumerate()
    .for_each(|(out_idx, _)| {
      results.push(
        products
          .clone()
          .into_iter()
          .filter(|item| {
            item.starts_with(&search_word
              .chars()
              .skip(0)
              .take(out_idx + 1)
              .collect::<String>())
          })
          .take(3)
          .collect(),
      );
    });
  return results;
}
